Missing required module 'XXX' when building project with cached dependencies

mollyIV opened the pull request:
Resolves https://github.com/tuist/tuist/issues/2023

### Short description 📝

This pull request solves an issue about missing required module 'XXX' when running unit tests with cached dependencies. 

### Solution 📦

After calling `tuist focus FrameworkC -n` (step 3. from steps to repro) and opening a `FrameworkC.xcworkspace`, the project was not building with the following error:

```
// FrameworkC.swift file

import Foundation
import FrameworkB // <--- Missing required module 'FrameworkA' ❌

public final class FrameworkC {
    public static func hello() {
        print("Hello, from your FrameworkC")
        let _ = FrameworkB.hello()
    }
}
```

It turned out that the transitive project dependency `FrameworkA` was not added to Framework Search Paths (only the direct dependency `FrameworkB` was added):

<img width="700" alt="module-001" src="https://user-images.githubusercontent.com/10795657/99418152-01fa1100-28fb-11eb-9a39-a5b786c4bae5.png">

The fix is adding transitive dependencies of pre-compiled frameworks as linkable dependencies ✨ 

Thanks a million to @pepibumur for helping and guiding in investigation.

> More description will be delivered

### Testing ⚙️

**Acceptance tests**

I added a new acceptance test `ios_app_with_transitive_project`

```
Scenario: The project is an iOS application with a target dependency and transitive framework dependency (ios_app_with_transitive_project)
    Given that tuist is available
    And I have a working directory
    Then I copy the fixture ios_app_with_transitive_project into the working directory
    Then tuist warms the cache
    When I remove derived data directory
    When tuist focuses the target App
    Then I should be able to build for iOS the scheme App
```

**Manual testing**

I used a [demo project ](https://github.com/tuist/tuist/files/5523617/demo.zip) and  followed the repro steps. The transitive project dependency `FrameworkA` was added to Framework Search Paths:

<img width="700" alt="module-002" src="https://user-images.githubusercontent.com/10795657/99418452-52716e80-28fb-11eb-8bc4-0401efa490fe.png">

mollyIV commented on Sources/TuistCore/Graph/Graph.swift:
```diff
@@ -240,7 +240,7 @@ public class Graph: Encodable, Equatable {
         // Precompiled libraries and frameworks
 
-        let precompiledLibrariesAndFrameworks = targetNode.precompiledDependencies
+        let precompiledLibrariesAndFrameworks = targetNode.recursivePrecompiledDependencies
```
I was wondering if we shall filter only dynamic transitive dependencies and the direct dependencies of a target. 

Thoughts? 🤔

pepibumur replied:
I don't think this is right. Your change is saying that we should link both, direct and transitive dependencies. That's not what users would expect.

mollyIV replied:
If we didn't link the transitive dependencies of `.framework`s in case of precompiled dependencies, we would get an error about missing frameworks. It is shown on the error screenshot in the issue description. Frameworks cannot embed other frameworks, so just adding already pre-built frameworks in `.tuist/Cache` to Framework Search Paths seemed to be a good idea 🤔

pepibumur commented on Sources/TuistCore/Graph/Nodes/TargetNode.swift:
```diff
@@ -113,3 +114,15 @@ public class TargetNode: GraphNode {
     }
 }
+
+private extension PrecompiledNode {
```
I think this can be useful in other places so I'd rather move it to the body of `PrecompiledNode`.

mollyIV replied:
I got rid of this property, because it is not needed anymore.

pepibumur commented on Sources/TuistCore/Graph/Nodes/TargetNode.swift:
```diff
@@ -113,3 +114,15 @@ public class TargetNode: GraphNode {
+    }
+}
+
+private func recursiveNodeDependencies(_ node: PrecompiledNode) -> [PrecompiledNode] {
```
We have a method in `Graph` for traversing the graph making sure there's only one pass through each of the nodes.

mollyIV replied:
Oh, didn't see it 👍 Assume we are talking about `Graph.findAll` method, right?

pepibumur commented on Sources/TuistCore/Graph/Nodes/TargetNode.swift:
```diff
@@ -88,8 +88,9 @@ public class TargetNode: GraphNode {
 
-    public var precompiledDependencies: [PrecompiledNode] {
-        dependencies.lazy.compactMap { $0 as? PrecompiledNode }
+    public var recursivePrecompiledDependencies: [PrecompiledNode] {
```
As mentioned below, I think traversing a graph is knowledge that should live in the graph and not in the nodes.

mollyIV replied:
I got rid of this method in favour of `findAll(...)` in a `Graph` 👍

pepibumur reviewed (commented):
@mollyIV I'm a bit confused. Are you using static or dynamic frameworks? Assuming the following dependency graph:


```
B -> C -> A
```

SINCE B doesn't link A and I don't think its directory doesn't have to be exposed through `FRAMEWORK_SEARCH_PATHS` *(unless I'm missing something)*. However, its directory needs to be exposed to the dynamic linker through the `LD_RUNPATH_SEARCH_PATHS` build setting. Have you tried to do the latter instead? 

@kwridan & @andreacipriani might be able to help to understand this problem.

mollyIV commented:
Hey @pepibumur 👋 

Thanks for a quick response. 

> Assuming the following dependency graph:

Let me clarify the demo project. The project consists of an iOS application. The app links `FrameworkC` directly. The `FrameworkC` links `FrameworkB` directly. The `FrameworkB` links `FrameworkA` directly. 

```
App <-- FrameworkC <-- FrameworkB <-- FrameworkA
```

> Are you using static or dynamic frameworks?

I am warming up the cache via `tuist cache warm` command, so all the dependencies of the app are `.framework`s.

> SINCE B doesn't link A

So this ☝️ isn't right, `FrameworkB` links `FrameworkA`:

```
// Project.swift

import ProjectDescription
import ProjectDescriptionHelpers

let project = Project.framework(
    name: "FrameworkB",
    dependencies: [
        .project(
            target: "FrameworkA",
            path: .relativeToRoot("Projects/FrameworkA")
        )
    ]
)
```

mollyIV commented:
Hey @pepibumur 👋 

I added an acceptance test with a new fixture. I hope it's gonna help to clarify the confusions. Cheers!

pepibumur commented on Sources/TuistCore/Graph/Graph.swift:
```diff
@@ -240,9 +240,12 @@ public class Graph: Encodable, Equatable {
-        let precompiledLibrariesAndFrameworks = targetNode.precompiledDependencies
-            .lazy
-            .map(GraphDependencyReference.init)
+        let precompiledLibrariesAndFrameworks = findAll(
```
I think I have a better idea of why this change is necessary. If I'm not wrong, the issue is that transitive pre-compiled dependencies are not linked, right? I think we might need to adjust your changes a bit further. We might have the following scenarios:

1. App ---(depends on)---> Precompiled static binary (A) ---> Precompiled static binary (B)
2. App ---(depends on)---> Precompiled dynamic binary (A) ----> Precompiled dynamic binary (B)
3. App ---(depends on)---> Precompiled static binary (A) ----> Precompiled dynamic binary (B)

I think what we should get at the app level is:
1. App links both binaries
2. App links A, and copies A and B into the product.
3. App links A, and copies B into the product.

If I'm not wrong, with your changes in the 3rd case App would be linking B, which is not what we want because that'll create an implicit dependency between the app and B and users will be able to import B.

Notice that `findAll` has 2 arguments, `test`, and `skip` that you can use to select the right dependencies here:

```swift
findAll(targetNode: targetNode, test: and(isStatic, isPrecompiled), skip: canLinkStaticProducts)
```

@kwridan am I missing something? 

I think we didn't have this scenario before caching because we didn't have the notion of dependencies between pre-compiled nodes. However, that's a thing now and we should account for it.

kwridan replied:
> I think we didn't have this scenario before caching because we didn't have the notion of dependencies between pre-compiled nodes. However, that's a thing now and we should account for it.

Indeed, precompiled frameworks depending on one another is an internal detail used for `tuist cache` and not something exposed during non-cached workflows.

To cater for the scenarios above perhaps something less eager is needed. I believe @mollyIV first attempt at this was more correct:

https://github.com/tuist/tuist/pull/2051/commits/c317a709a30dfdc7cd55b459c12bb47b5c048d36

The latest change here will transitively include ALL transitive precompiled nodes:

- Target A > Precompiled A
- Target A > Target B > Precompiled B

This change would result in TargetA ending up with `Precompiled A` and `Precompiled B` which is not the correct behaviour in all circumstances (we have that covered [in the next set of lines below](https://github.com/tuist/tuist/blob/main/Sources/TuistCore/Graph/Graph.swift#L249))


We can see the impact/change in behaviour in the fixture `ios_app_with_transitive_framework`: 
- generate the projects (without caching) using latest tuist and this PR :/ 

This PR:

![Screenshot 2020-12-07 at 10 12 22](https://user-images.githubusercontent.com/11914919/101338323-b93bd500-3874-11eb-97fa-ca48fa9d201b.png)

![Screenshot 2020-12-07 at 10 09 34](https://user-images.githubusercontent.com/11914919/101338247-9f01f700-3874-11eb-8e2e-2929f10bf933.png)

![Screenshot 2020-12-07 at 10 09 27](https://user-images.githubusercontent.com/11914919/101338242-9e696080-3874-11eb-83c2-9f962b2500cf.png)

Latest tuist release _(correct behaviour)_:

![Screenshot 2020-12-07 at 10 10 02](https://user-images.githubusercontent.com/11914919/101338396-d2dd1c80-3874-11eb-9ff6-2bbb004a1c71.png)

![Screenshot 2020-12-07 at 10 13 41](https://user-images.githubusercontent.com/11914919/101338483-f011eb00-3874-11eb-8fd7-baa4928be3a0.png)

![Screenshot 2020-12-07 at 10 13 51](https://user-images.githubusercontent.com/11914919/101338500-f607cc00-3874-11eb-8b13-b71876f1da39.png)

kwridan replied:
Follow up from slack discussions:

- The recursive precompiled dependencies may need to be augmented with some additional logic as suggested by @pepibumur 

We'll need to ensure we traverse the `PrecompiledNode`s path as opposed the `TargetNode`s - its a little tricky as the `findAll` was geared towards `TargetNode` traversal.

![precompiled-nodes](https://user-images.githubusercontent.com/11914919/101348133-0f177980-3883-11eb-914f-769bc70ded11.png)


Something like this may do the trick:


```swift
let isPrecompiled: (GraphNode) -> Bool = { $0 is PrecompiledNode }
let isDynamicPrecompiledNode: (PrecompiledNode) -> Bool = { $0.isDynamicAndLinkable() }
let isTargetNode: (GraphNode) -> Bool = { $0 is TargetNode }
let precompiledLibrariesAndFrameworks = findAll(targetNode: targetNode, test: isPrecompiled, skip: or(isDynamicPrecompiledNode, isTargetNode))
        .lazy
        .map(GraphDependencyReference.init)
```

Yes in some cases we do need to care about target transitive precompiled nodes, but that [logic is already in place](https://github.com/tuist/tuist/blob/main/Sources/TuistCore/Graph/Graph.swift#L249) :)

We can try to formalise those into unit tests within the `Graph` and `ValueGraphTraverser`

Scenarios:

**1. App ---(depends on)---> Precompiled static binary (A) ---> Precompiled static binary (B)**

Linkable: 
- Precompiled static binary (A)
- Precompiled static binary (B)

Embeddable: _None_

**2. App ---(depends on)---> Precompiled dynamic binary (A) ----> Precompiled dynamic binary (B)**

Linkable: 
- Precompiled dynamic binary (A)

Embeddable:
- Precompiled dynamic binary (A)
- Precompiled dynamic binary (B)

**3. App ---(depends on)---> Precompiled static binary (A) ----> Precompiled dynamic binary (B)**

Linkable: 
- Precompiled static binary (A)
- Precompiled dynamic binary (B)

Embeddable:
- Precompiled dynamic binary (B)

**4. App ---(depends on)---> Precompiled dynamic binary (A) ----> Precompiled static binary (B)**

Linkable: 
- Precompiled dynamic binary (A)

Embeddable:
- Precompiled dynamic binary (A)

Notes:
- In this case Precompiled static binary (B) has already been linked into Precompiled dynamic binary (A) - as such no further action is needed

**5. App ---(depends on)---> Precompiled static binary (A)**

Linkable: 
- Precompiled static binary (A)

Embeddable: None

**6. App ---(depends on)---> Precompiled dynamic binary (A)**

Linkable: 
- Precompiled dynamic binary (A)

Embeddable:
- Precompiled dynamic binary (A)

pepibumur reviewed (commented):
Thank you very much for adding a fixture @mollyIV. I left a comment regarding the approach you are taking. Also, once we find out what's the right solution for this, we'll have to update the logic in the `ValueGraph` too because we are deprecating `Graph`.

kwridan reviewed (commented):
thanks @mollyIV and @pepibumur - reasoning about this can be quite tricky 😅

I believe the first attempt with `recursivePrecompiledFrameworks` was along the correct lines perhaps needs a few more tests cases / fixtures to ensure we handle the scenarios @pepibumur described are handled gracefully.

That wouldn't impact non-cached workflows and caters for the scenarios where the newly cached (precompiled) target has a dependency on a precompiled framework.

kwridan reviewed (commented):
Was debugging this locally with @mollyIV  - looks like @mollyIV first solution is the most suitable one with minimal impact on non cached workflows. 

Seeing `Precompiled` > `Precompiled` dependencies isn't a publicly supported dependency we don't need to address all the edge cases of it at the moment - it's only used internally as a `tuist cache` `tuist focus` implementation detail - we can refine it as we better under stand it  :) 

Thanks for everyone's efforts on this! and thanks for your patience @mollyIV 👍

mollyIV commented:
Thanks for your valuable input @kwridan and @pepibumur 🙇 

As agreed on, I reverted the implementation to the first commit. I added the suggested unit tests to a `Graph`. However, I changed the expectations for #2:

```
// App ---(depends on)---> Precompiled dynamic binary (A) ----> Precompiled dynamic binary (B)
```

I did a quick test and the compiler wasn't happy when there was no linking of precompiled dynamic binary (B) in the app target, even though a `Framewok2` was embedded into the `App`:

```
Missing required module 'Framework2'
```

Haven't tested the #3 yet, but according to unit tests, the same issue may probably be there. 

Thoughts? 🤔

kwridan commented:
Thanks @mollyIV - indeed it seems our original expectations for scenario 2 don't quite work with the same precompiled libraries as they do with regular targets. The changes you made to the expectations are fine, they are only scoped to the precompiled nodes.

Looks like you'll need to rebase your branch 🤐 all the `GraphTests` have been migrated over to `ValueGraphTraverserTests` - the idea there is to have the logic applied to both the reference `Graph` and the new `ValueGraph` and verifying the results from both graphs are identical to ensure we can switch to the new `ValueGraph` smoothly. For the `ValueGraph` all the traversal logic now lives independently in `ValueGraphTraverser`.

mollyIV commented:
> Looks like you'll need to rebase your branch 🤐

Will do, thanks for the context. I am going to mirror the changes related to precompiled binaries to `ValueGraphTraverser ` and `ValueGraphTraverserTests`.

mollyIV commented on Tests/TuistCoreTests/ValueGraph/ValueGraphTraverserTests.swift:
```diff
@@ -1525,6 +1525,58 @@ final class ValueGraphTraverserTests: TuistUnitTestCase {
         XCTAssertEqual(gotGraph.first, GraphDependencyReference(precompiledNode: precompiledNode))
     }
+    
+    func test_linkableAndEmbeddableDependencies_when_appDependensOnPrecompiledStaticBinaryWithPrecompiledStaticBinaryDependency() throws {
```
I'd like to kindly ask you @kwridan to check if the "Given: Value Graph" part of a unit test is correct 🙏  Please note that this test is failing, because a value graph linkable output is not equal to graph traverser output. 

```
XCTAssertEqual(got, gotGraph)
```

I haven't made the changes in the `ValueGraph` yet, however, I am wondering if maybe we should deliver it in the separate pull request due to taking time to figure out the best solution and the fact that we already have pretty decent solution for a reference graph. What do you think? CC: @pepibumur

mollyIV commented:
Hey @kwridan 👋 I implemented the suggested changes, kindly ask for second round of code review, thanks! 😊

kwridan reviewed (approved):
Nice work @mollyIV! looks like you've covered all the cases 🎉

mollyIV commented on Tests/TuistCoreTests/ValueGraph/ValueGraphTraverserTests.swift:
```diff
@@ -1526,6 +1526,315 @@ final class ValueGraphTraverserTests: TuistUnitTestCase {
+        XCTAssertEqual(gotGraph.first, GraphDependencyReference(precompiledNode: precompiledStaticBinaryA))
+        XCTAssertEqual(gotGraph.last, GraphDependencyReference(precompiledNode: precompiledStaticBinaryB))
+        
+        // When
+        let embeddable = graphTraverser.embeddableFrameworks(path: project.path, name: target.name)
+        
+        // Then
+        XCTAssertTrue(embeddable.isEmpty)
```
Good point! Let me add these 💪

mollyIV replied:
Done 🔥

pepibumur commented:
Solid work here @mollyIV and @kwridan 👏 👏. Extending our traversing logic to account for transitive pre-compiled nodes is crucial for the cache mutation to work as expected.

![](https://media2.giphy.com/media/6q0XwC1PfHRrZOYeah/giphy.gif?cid=5a38a5a2x57yt9akh6dbiln0j44ahz3ei17h2oe9b2jlhr4l&rid=giphy.gif)

mollyIV commented on Tests/TuistCoreTests/ValueGraph/ValueGraphTraverserTests.swift:
```diff
@@ -1525,6 +1525,58 @@ final class ValueGraphTraverserTests: TuistUnitTestCase {
+        // Then
+        XCTAssertEqual(got, gotGraph)
+        XCTAssertEqual(gotGraph.first, GraphDependencyReference(precompiledNode: precompiledStaticBinaryA))
+        XCTAssertEqual(gotGraph.last, GraphDependencyReference(precompiledNode: precompiledStaticBinaryB))
```
Done 😊

mollyIV commented:
Thank you sooo much for all the help Pedro and Kas! Let's :shipit:

pepibumur commented:
I fixed the linting issues after the merge :)
